File: day08/cpp/main.cpp
#include <fstream>
#include <iostream>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

int main(int argc, char *argv[]) {
  fstream fs(argv[1]);
  vector<string> grid;
  for (string line; fs >> line; grid.push_back(line))
    ;
  unordered_map<char, vector<pair<int, int>>> antennas;
  for (size_t i{0}; i < grid.size(); i++) {
    for (size_t j{0}; j < grid[i].size(); j++) {
      if (grid[i][j] != '.') {
        antennas[grid[i][j]].push_back({i, j});
      }
    }
  }

  constexpr auto hash_pair = [](const pair<int, int> &v) -> int {
    return v.first * 130 + v.second;
  };

  unordered_set<pair<int, int>, decltype(hash_pair)> antinodes;

  auto in_bounds = [&](int i, int j) -> bool {
    return i >= 0 && i < grid.size() && j >= 0 && j < grid[i].size();
  };

  for (auto [f, p] : antennas) {
    for (int i = 0; i < p.size(); i++) {
      for (int j = 0; j < p.size(); j++) {
        if (i == j) {
          continue;
        }

        pair<int, int> itoj = {p[j].first - p[i].first,
                               p[j].second - p[i].second};
        int pi = p[i].first;
        int pj = p[i].second;

        while (in_bounds(pi, pj)) {
          antinodes.insert({pi, pj});
          pi = pi + itoj.first;
          pj = pj + itoj.second;
        }
      }
    }
  }

  cout << antinodes.size() << endl;
}

File: day08/dart/main.dart
import 'dart:collection';
import 'dart:io';

void main(List<String> args) async {
  final file = File(args.first);
  final grid = await file.readAsLines();

  final antennas = HashMap<String, List<(int, int)>>();
  for (int i = 0; i < grid.length; i++) {
    for (int j = 0; j < grid[i].length; j++) {
      String c = grid[i][j];
      if (c != '.') {
        if (!antennas.containsKey(c)) antennas[c] = [];
        antennas[c]!.add((i, j));
      }
    }
  }

  final inBounds = (int x, int y) {
    return x >= 0 && x < grid.length && y >= 0 && y < grid[x].length;
  };

  final antinodes = HashSet<(int, int)>();
  for (final item in antennas.entries) {
    final pos = item.value;
    for (int i = 0; i < pos.length; i++) {
      for (int j = 0; j < pos.length; j++) {
        if (i == j) {
          continue;
        }

        final itoj = (pos[j].$1 - pos[i].$1, pos[j].$2 - pos[i].$2);
        int x = pos[i].$1;
        int y = pos[i].$2;

        while (inBounds(x, y)) {
          antinodes.add((x, y));
          x += itoj.$1;
          y += itoj.$2;
        }
      }
    }
  }

  print(antinodes.length);
}
